### PR TITLE
service.go: fix repair semantics regression

### LINKED ISSUE #3463: Too high repair parallelism
Version: 3.1.1
Cluster size: 9
RF: 3
parallel: 0
max_repair_ranges_in_parallel: 6
intensity: 1

*Description*
According to SM doc pages (https://manager.docs.scylladb.com/stable/sctool/repair.html#id8) the parallelism for a KS with RF=3 and cluster with 9 nodes is supposed to be 3.

However we see that the effective parallelism is much more than that:

```
$ run scylla_hosts curl  http://127.0.0.1:10000/storage_service/active_repair
Elapsed: 00:01 - running on Scylla nodes: 'curl http://127.0.0.1:10000/storage_service/active_repair'...-
running on Scylla nodes: 'curl http://127.0.0.1:10000/storage_service/active_repair': done. Took: 1.43 seconds.
╒══════════════════════════════════════════╤═══════════════════════════════════════╤══════════════╕
│ Scylla host                              │ Result                                │ Duration     │
╞══════════════════════════════════════════╪═══════════════════════════════════════╪══════════════╡
│ scylla-xxx-xxxx-xxxx-asia-southeast1-a-1 │ []                                    │ 0.68 seconds │
├──────────────────────────────────────────┼───────────────────────────────────────┼──────────────┤
│ scylla-xxx-xxxx-xxxx-asia-southeast1-b-1 │ [5,4,3,2,1]                           │ 0.67 seconds │
├──────────────────────────────────────────┼───────────────────────────────────────┼──────────────┤
│ scylla-xxx-xxxx-xxxx-asia-southeast1-c-1 │ [14,13,12,11,10,9,8,7,6]              │ 0.67 seconds │
├──────────────────────────────────────────┼───────────────────────────────────────┼──────────────┤
│ scylla-xxx-xxxx-xxxx-asia-southeast1-a-3 │ [5,4,3,2,1]                           │ 0.68 seconds │
├──────────────────────────────────────────┼───────────────────────────────────────┼──────────────┤
│ scylla-xxx-xxxx-xxxx-asia-southeast1-b-3 │ []                                    │ 0.68 seconds │
├──────────────────────────────────────────┼───────────────────────────────────────┼──────────────┤
│ scylla-xxx-xxxx-xxxx-asia-southeast1-c-3 │ [5,4,3,2,1]                           │ 0.69 seconds │
├──────────────────────────────────────────┼───────────────────────────────────────┼──────────────┤
│ scylla-xxx-xxxx-xxxx-asia-southeast1-a-4 │ []                                    │ 0.68 seconds │
├──────────────────────────────────────────┼───────────────────────────────────────┼──────────────┤
│ scylla-xxx-xxxx-xxxx-asia-southeast1-b-4 │ [151,150,149,148,147,146,145,144,143] │ 0.68 seconds │
├──────────────────────────────────────────┼───────────────────────────────────────┼──────────────┤
│ scylla-xxx-xxxx-xxxx-asia-southeast1-c-4 │ [37,36,35,33,34]                      │ 0.67 seconds │
╘══════════════════════════════════════════╧═══════════════════════════════════════╧══════════════╛
```

As we can see there are 4*5 + 2*9 = 38 repair jobs running at the same time. 
This doesn't seem to correlate with the documentation referenced above.

### PR DESCRIPTION
**Pull request by vladzcloudius**
Restore the repair task sanity.

Make the code correspond design and Documentation.
See more in the patches description.

Fixes #3463 and #3424

### CODE REVIEW

**Comment by Michal-Leszczynski** [comment]
It doesn't look related to #3464.

Moreover, we this patch won't really work the way you intended.
Current implementation uses repair controller in order to check if given job repair should be scheduled. Basically, repair worker asks if it can repair `intensity` amount of token ranges and controller checks if it fits under the limits (that you changed). So when `intensity > 1`, this will always fail.
On the other hand, controller has some additional logic, which always allows to schedule a repair job if there are no other repair jobs running in the cluster and that's why the tests have passed (but this means that there is only 1 repair job per cluster running, so repair will be really slow).

A quick patch for this was introduced to branch 3.1 via #3496.

**Comment by vladzcloudius** [comment]
> It doesn't look related to #3464.

Right. I meant #3463 of course.
Thanks for looking at it, @Michal-Leszczynski,  and, to clarify, I didn't mean this to be merged as v1.
This is my "toy" Go attempt. So, I'm sure I'm going to need a lot of help here.
More below.

> 
> Moreover, we this patch won't really work the way you intended. Current implementation uses repair controller in order to check if given job repair should be scheduled. Basically, repair worker asks if it can repair `intensity` amount of token ranges and controller checks if it fits under the limits (that you changed). So when `intensity > 1`, this will always fail. 

I'm well aware of what you have described above and this was exactly my intention in this patch (you could see that from the inline comment).

However in following updates I'm going to bring back an integer semantics of `intensity` and restore the `parallel` semantics and requirements in full.

In particular the current ` rowLevelRepairController::shouldBlock(...)` has an outdated logic that is based on the old design where a per-shard range repair was an atomic repair item. And hence all the accounting.

Let's fix it here. More below.

> On the other hand, controller has some additional logic, which always allows to schedule a repair job if there are no other repair jobs running in the cluster and that's why the tests have passed (but this means that there is only 1 repair job per cluster running, so repair will be really slow).

Right. The logic that allocates workers.
Let's fix that.

> 
> A quick patch for this was introduced to branch 3.1 via #3496.

That patch doesn't help us much because it seems rather bogus to me since it still mixes and confuses "token ranges" and "token tasks". 
Specifically I suspect that with that patch there would still be num_shards token ranges repaired in a single repair task.

**Comment by karol-kokoszka** [comment]
@vladzcloudius the PR ensuring single (or predefined number) job to be executed is already merged to branch-3.1.
What do you want to achieve with that PR ? I'm not gonna merge that.

**Comment by vladzcloudius** [comment]
> @vladzcloudius the PR ensuring single (or predefined number) job to be executed is already merged to branch-3.1. What do you want to achieve with that PR ? I'm not gonna merge that.

@karol-kokoszka The PR you are referring is not solving the problem in question which is: restore the documented SM repair functionality and constraints.

Your PR does something else - something only you understand.

We are going to have a call with Tzach first thing next week and me and Asias are going to insist on reverting your PR. See more details on corresponding GH issues.

As to this PR - this is a demonstration how the documented functionality can be restored. 

cc @mykaul @harel-z @ruthea

**Comment by vladzcloudius** [comment]
@asias take a glimpse. This PR is supposed to implement your proposal from the "Repair 2023 Design".

The only thing that is missing is the way to fetch `max_repair_ranges_in_parallel` values from nodes.
AFAICT today it's not exposed anywhere except for a syslog, right?
If "yes", then I'll open a GH issue and send a `scylla` PR for that adds a corresponding REST API.

@eliransin you probably want to keep and eye on this as well.

**Comment by asias** [comment]
> @asias take a glimpse. This PR is supposed to implement your proposal from the "Repair 2023 Design".
> 
> The only thing that is missing is the way to fetch `max_repair_ranges_in_parallel` values from nodes. AFAICT today it's not exposed anywhere except for a syslog, right? If "yes", then I'll open a GH issue and send a `scylla` PR for that adds a corresponding REST API.

IIRC, SM calculates max_repair_ranges_in_parallel directly using the same method as in scylla core:

auto max_memory_repair = memory::stats().total_memory() * 0.1; // 10% of per shard memory size
max_repair_memory_per_range = 32 * 1024 * 1024;
max_repair_ranges_in_parallel = max_repair_memory / max_repair_memory_per_range   / 4;

We could expose with api as well. 

> @eliransin you probably want to keep and eye on this as well.

**Comment by vladzcloudius** [comment]
> > @asias take a glimpse. This PR is supposed to implement your proposal from the "Repair 2023 Design".
> > The only thing that is missing is the way to fetch `max_repair_ranges_in_parallel` values from nodes. AFAICT today it's not exposed anywhere except for a syslog, right? If "yes", then I'll open a GH issue and send a `scylla` PR for that adds a corresponding REST API.
> 
> IIRC, SM calculates max_repair_ranges_in_parallel directly using the same method as in scylla core:
> 
> auto max_memory_repair = memory::stats().total_memory() * 0.1; // 10% of per shard memory size max_repair_memory_per_range = 32 * 1024 * 1024; max_repair_ranges_in_parallel = max_repair_memory / max_repair_memory_per_range / 4;
> 
> We could expose with api as well.
> 
> > @eliransin you probably want to keep and eye on this as well.

Correct, you can see how I nucked the corresponding call in this PR. I didn't like it. But maybe I should bring it back for now. Still better than a hard coded value...;) 
I think that there must be a single "source of thruth" for this value and it must be scylla.
If tomorrow you change scylla logic around it it must be transparent to SM.

File: pkg/service/repair/service.go
Comment (tarzanek) [inline pkg/service/repair/service.go:592]: revert and use above function, as per Asias this is OK
(it can be reverted to REST call later, so TODO can be moved here)

File: pkg/service/repair/controller.go
Comment (tarzanek) [inline pkg/service/repair/controller.go:209]: you stop considering busy replicas here
isn't parallel a static var?

Comment (vladzcloudius) [reply pkg/service/repair/controller.go:209]: It turns out that this (and the condition above) was a total BS.
And one simply needs to compare `c.allJobs` (counter of currently running repair tasks) vs `c.intensity.Parallel()` - the actual value of `--parallel` for this repair task.

**Comment by vladzcloudius** [comment]
@tarzanek @igorribeiroduarte @asias please, re-review. 
This PR is meant to be merged: I rebased and split it into meaningful patches.

**Comment by tzach** [comment]
I'm worried about this patch updating the default behavior of *all* customers. 
This is something we try not to do in a patch release.  For sure, it makes sense for some customers, but I still need to get a blessing from QA and @asias it makes sense in *all* cases.

I agree with @vladzcloudius we have a difference between docs and logic we need to fix, but the issue has been there for years and works reasonably for most customers. We need to validate we will not harm a subset of the deployments.

File: pkg/service/repair/controller.go
Comment (tarzanek) [inline pkg/service/repair/controller.go:216]: instead of comments isn't there some debug flag or logger (debug) level where you can put this?

**Comment by tarzanek** [comment]
@tzach it should of course be tested, but I don't expect regression, but rather performance improvement of repairs for all customers (since right now it's unpredictable for clusters with more nodes - if you are lucky, your repair won't block with scheduling parallel tasks on same node and it will be done quickly
if you are unlucky, you repair takes a long time due to racing for resources by same shards for multiple repair tasks on same node)
- repair time should get more predictable once this *bug fix* is merged - and we should avoid seeing repair "stuck" for weird reasons
(I wrote bug fix and consider this as a bug fix - since the behaviour IS broken now, it breaks the simple rule which is that 1 node runs only 1 parallel task - and @asias will bless this)

so QA should mainly test this patch in positive sense on e.g. 3 node clusters with lots of cpus (>16 ?) per node and of course classic positive 3 node cluster with low amount of cpus per node

and negative test cases - multi node clusters (3, 4 nodes per rack) again with more cpus
and
Vlad found it on a certain customer setup and it also could be taken as a model (so on top of above there are MVs to make the repair more funny)

**Comment by mykaul** [comment]
> @tzach it should of course be tested, but I don't expect regression, but rather performance improvement of repairs for all customers (since right now it's unpredictable for clusters with more nodes - if you are lucky, your repair won't block with scheduling parallel tasks on same node and it will be done quickly if you are unlucky, you repair takes a long time due to racing for resources by same shards for multiple repair tasks on same node) repair time should get more predictable once this _bug fix_ is merged - and we should avoid seeing repair "stuck" for weird reasons (I wrote bug fix and consider this as a bug fix - since the behaviour IS broken now, it breaks the simple rule which is that 1 node runs only 1 parallel task - and @asias will bless this)
> 
> so QA should mainly test this patch in positive sense on e.g. 3 node clusters with lots of cpus (>16 ?) per node and of course classic positive 3 node cluster with low amount of cpus per node

Ideally, tests should be added to the PR. To begin with.

File: pkg/service/repair/controller.go
Comment (vladzcloudius) [reply pkg/service/repair/controller.go:216]: I didn't want to create too much "noise" in this PR.
I was planning to do this in the followup patches.
If it bothers you I can drop all this patch.

**Comment by vladzcloudius** [comment]
> > @tzach it should of course be tested, but I don't expect regression, but rather performance improvement of repairs for all customers (since right now it's unpredictable for clusters with more nodes - if you are lucky, your repair won't block with scheduling parallel tasks on same node and it will be done quickly if you are unlucky, you repair takes a long time due to racing for resources by same shards for multiple repair tasks on same node) repair time should get more predictable once this _bug fix_ is merged - and we should avoid seeing repair "stuck" for weird reasons (I wrote bug fix and consider this as a bug fix - since the behaviour IS broken now, it breaks the simple rule which is that 1 node runs only 1 parallel task - and @asias will bless this)
> > so QA should mainly test this patch in positive sense on e.g. 3 node clusters with lots of cpus (>16 ?) per node and of course classic positive 3 node cluster with low amount of cpus per node
> 
> Ideally, tests should be added to the PR. To begin with.

I've tested this PR ma

> I'm worried about this patch updating the default behavior of _all_ customers. This is something we try not to do in a patch release. For sure, it makes sense for some customers, but I still need to get a blessing from QA and @asias it makes sense in _all_ cases.

What are you talking about, @tzach?

> 
> I agree with @vladzcloudius we have a difference between docs and logic we need to fix, but the issue has been there for years and works reasonably for most customers. 

No it didn't, Tzach. We have customers complaining about "too aggressive repairs" that impact tail latencies all the time.

Please, check the amount of corresponding GH issues on the matter.

> We need to validate we will not harm a subset of the deployments.

Could you, please, be more specific about what you think can harm and what?

**Comment by vladzcloudius** [comment]
> > @tzach it should of course be tested, but I don't expect regression, but rather performance improvement of repairs for all customers (since right now it's unpredictable for clusters with more nodes - if you are lucky, your repair won't block with scheduling parallel tasks on same node and it will be done quickly if you are unlucky, you repair takes a long time due to racing for resources by same shards for multiple repair tasks on same node) repair time should get more predictable once this _bug fix_ is merged - and we should avoid seeing repair "stuck" for weird reasons (I wrote bug fix and consider this as a bug fix - since the behaviour IS broken now, it breaks the simple rule which is that 1 node runs only 1 parallel task - and @asias will bless this)
> > so QA should mainly test this patch in positive sense on e.g. 3 node clusters with lots of cpus (>16 ?) per node and of course classic positive 3 node cluster with low amount of cpus per node
> 
> Ideally, tests should be added to the PR. To begin with.

Ideally R&D would be fixing it and not me, @mykaul.

I've tested it manually quite extensively and there are quite a lot of CI tests maintained by our SM mainteners. If some new tests need to be added - we have the whole QA and SM teams for that. I'm not familiar with the SM testing framework but I can definitely explain how the corresponding test can be defined, @mykaul.

To remind you, @mykaul @tzach, we need these issues to be fixed ASAP since we are bleeding in the field. And it was expected that R&D would do the fixing. However you know exactly what is going on with that.
So, I spent a few hours on my time on that to write and test this thing. However I expect the relevant R&D (!!!) members do what's left - like tests.
I really don't have too much bandwidth to learn your testing frameworks and work on tests.

**Comment by vladzcloudius** [comment]
> I'm worried about this patch updating the default behavior of _all_ customers. This is something we try not to do in a patch release. For sure, it makes sense for some customers, but I still need to get a blessing from QA and @asias it makes sense in _all_ cases.
> 
> I agree with @vladzcloudius we have a difference between docs and logic we need to fix, but the issue has been there for years and works reasonably for most customers. We need to validate we will not harm a subset of the deployments.

And most importantly, @tzach, what "patch release" you are referring here?
This is a PR towards the 'master'. You are welcome to backport it to whatever release you want.

**Comment by vladzcloudius** [comment]
@tzach @mykaul unless you have specific comments about this PR's code i suggest we merge it to the master and then your teams can test it, backport it or not.

I'll take a look on failed CI tests before that and if these failures are related to this PR i'll fix tests. 

Any objections?

**Comment by vladzcloudius** [comment]
For example, the same tests that were used/added in this (https://github.com/scylladb/scylla-manager/pull/3496) PR would cover the this PR as well, @mykaul

**Comment by vladzcloudius** [comment]
@karol-kokoszka could you tell me how I can either see the detailed output of the failed test of run the test myself?

**Comment by vladzcloudius** [comment]
@karol-kokoszka @Michal-Leszczynski could you, please, tell me how I can see the detailed output of the failed CI test(s) or (preferably) if I can run these tests locally?

**Comment by vladzcloudius** [comment]
FYI: the `make test` fails with a clean `master` with the following error:

```
ok      github.com/scylladb/scylla-manager/v3/pkg/util/tickrun  (cached)        coverage: 100.0% of statements
--- FAIL: TestTodayMidnight (0.00s)
    timeutc_test.go:13: invalid hour 2023-08-08 20:00:00 -0400 EDT
FAIL
```

And I don't see how it can possibly pass unless you run the above in the UTC timezone.

**Comment by Michal-Leszczynski** [comment]
Just so it's clear, SM devs are also preparing a fix for that (and many other repair issues #3423) and we plan to have implementation finished by the end of the week.

> @karol-kokoszka @Michal-Leszczynski could you, please, tell me how I can see the detailed output of the failed CI test(s) or (preferably) if I can run these tests locally?

I usually just go to actions, select your gh actions run, select failed test scenario, cogwheel, view raw logs and then I look for "--- FAIL" in the output. Running all the tests locally is just:
```
make start-dev-env
make check
make test
```
Testing only given pkg can be done by `make pkg-integration-test PKG=./pkg/service/repair`

**Comment by tzach** [comment]
> Just so it's clear, SM devs are also preparing a fix for that (and many other repair issues https://github.com/scylladb/scylla-manager/issues/3423) and we plan to have implementation finished by the end of the week.

Great!

Is the solution all in one PR / Commit?
It would be helpful to backport only the fix for #3463 to 3.1 to give a quick solution for user issue and keep the rest of the improvement to 3.2

File: pkg/service/repair/controller.go
Comment (mykaul) [inline pkg/service/repair/controller.go:6]: ?

File: pkg/service/repair/controller.go
Comment (mykaul) [inline pkg/service/repair/controller.go:195]: ?

File: pkg/service/repair/service.go
Comment (mykaul) [inline pkg/service/repair/service.go:575]:
```suggestion
			//TODO: collect max_repair_ranges_in_parallel from scylla hosts using the new API (see https://github.com/scylladb/scylladb/pull/14886 )
```

**Review by tarzanek — APPROVED** [review]
reconsider if fmt debug patch is needed, otherwise lgtm

**Comment by vladzcloudius** [comment]
> Just so it's clear, SM devs are also preparing a fix for that (and many other repair issues #3423) and we plan to have implementation finished by the end of the week.

Not sure what you are trying to say here, @Michal-Leszczynski.
This PR is fixing 2 issues out of those you referenced.
If you have comments on this PR you should state them.

Re-doing the same thing somebody already did is counterproductive in general.

> 
> > @karol-kokoszka @Michal-Leszczynski could you, please, tell me how I can see the detailed output of the failed CI test(s) or (preferably) if I can run these tests locally?
> 
> I usually just go to actions, select your gh actions run, select failed test scenario, cogwheel, view raw logs and then I look for "--- FAIL" in the output. Running all the tests locally is just:
> 
> ```
> make start-dev-env
> make check
> make test
> ```
> 
> Testing only given pkg can be done by `make pkg-integration-test PKG=./pkg/service/repair`

I'll try. Thanks.

File: pkg/service/repair/controller.go
Comment (vladzcloudius) [reply pkg/service/repair/controller.go]: resolved (I removed the last patch from this PR that adds those printouts).
I'll consider sending a followup with a proper logging.

File: pkg/service/repair/service.go
Comment (vladzcloudius) [reply pkg/service/repair/service.go:575]: Let me merge this.

File: pkg/service/repair/controller.go
Comment (vladzcloudius) [reply pkg/service/repair/controller.go]: I removed the last patch from the original PR that was adding these printouts.

File: pkg/service/repair/controller.go
Comment (asias) [inline pkg/service/repair/controller.go:102]: So here, if i == 0, we set it to max_repair_ranges_in_parallel. Otherwise, we set it to the number specified by the user.

File: pkg/service/repair/controller.go
Comment (asias) [inline pkg/service/repair/controller.go:193]: I am curious what was c.limits[h].Default?

File: pkg/service/repair/controller.go
Comment (asias) [inline pkg/service/repair/controller.go:201]: I assume c.allJobs tracks the number of pending jobs that are currently running. 

nit: We can use >= to avoid the plus one thing. Saying something in the comments like, we already have c.allJobs running and we can not have more than MaxParallel() jobs, so we need to block the request. 
```
if c.allJobs >= c.intensity.MaxParallel()  {
}
```

File: pkg/service/repair/controller.go
Comment (asias) [inline pkg/service/repair/controller.go:202]: Btw, it seems the function returns a reversed state. shouldBlock() returns false to perform the  block? Why?

File: pkg/service/repair/worker.go
Comment (asias) [inline pkg/service/repair/worker.go:82]: It is safe to drop the legacy partition level repair. All supported releases have row level repair now.

**Comment by asias** [comment]
I am not familiar with the SM code base, but the change looks reasonable to me. I have some comments inline. 

It is good that the number of changes is limited. 

```
$ git diff --stat 9b3527fc538381216ec7076a033f8d7363e91c6d
 pkg/service/repair/controller.go      | 92 +++++++++++++++++++++++++++++++++-----------------------------------------------------------
 pkg/service/repair/controller_test.go |  4 +---
 pkg/service/repair/service.go         | 20 ++++++++++++++------
 pkg/service/repair/worker.go          | 61 +------------------------------------------------------------
 4 files changed, 49 insertions(+), 128 deletions(-)
```

File: pkg/service/repair/service.go
Comment (vladzcloudius) [reply pkg/service/repair/service.go:574]: Great. That was exactly my idea. We are defaulting to `1` which is both to preserve the documented behavior and to be on a conservative side by default - this can be seen as a "least noisy fastest mode (since `parallel` defaults to 0)".

And then we can change the intensity all the way up to Max (I tested that it works manually).

File: pkg/service/repair/controller.go
Comment (vladzcloudius) [reply pkg/service/repair/controller.go:202]: This wasn't my decision - I'm just fixing what was written before me ;)
But I agree, some functions names do not match their semantics in a straight forward way.

But logic wise the code does what it's supposed to: https://github.com/scylladb/scylla-manager/blob/master/pkg/service/repair/controller.go#L49

File: pkg/service/repair/controller.go
Comment (vladzcloudius) [reply pkg/service/repair/controller.go:201]: Correct the result would be the same.
However I followed the previous code approach on this and wanted to keep different condition blocks (some of when remained) to look similar.

And hence the +1 thingy.
In practice, this is not a real-time application anyway so it's not about the performance to begin with.

And I kinda think that `if current_counter+1> max_counter; then {don't do something} else {++current_counter;}` is more readable than `>=` option - it semantically hints that we are checking a condition that if held is going to cause the `current_counter` to be incremented - thereby self-explaining its (`current_counter`) semantics.

File: pkg/service/repair/controller.go
Comment (vladzcloudius) [reply pkg/service/repair/controller.go:193]: AFAIR it's initialized in this hunk: https://github.com/scylladb/scylla-manager/pull/3506#discussion_r1289704818

So, technically, before this PR we allowed up to "number of shards" concurrent repair jobs in a single replica (other conditions aside).

Comment (vladzcloudius) [reply pkg/service/repair/controller.go:193]: AFAIR it's initialized in this hunk: https://github.com/scylladb/scylla-manager/pull/3506#discussion_r1289704818

So, technically, before this PR we allowed up to "number of shards" concurrent repair jobs in a single replica (other conditions aside).

This was the central part of the "bug" - SM code was mixing the "repair ranges math" with the "repair jobs" math.

It most probably came from the original repair code that was per-shard-ranges based. And probably it made sense then. But for a row-level repair things can and should be significantly simplified.

File: pkg/service/repair/controller.go
Comment (vladzcloudius) [reply pkg/service/repair/controller.go:102]: correct

File: pkg/service/repair/controller.go
Comment (vladzcloudius) [reply pkg/service/repair/controller.go:178]: correct again!

**Comment by vladzcloudius** [comment]
> I am not familiar with the SM code base, but the change looks reasonable to me. I have some comments inline.
> 
> It is good that the number of changes is limited.
> 
> ```
> $ git diff --stat 9b3527fc538381216ec7076a033f8d7363e91c6d
>  pkg/service/repair/controller.go      | 92 +++++++++++++++++++++++++++++++++-----------------------------------------------------------
>  pkg/service/repair/controller_test.go |  4 +---
>  pkg/service/repair/service.go         | 20 ++++++++++++++------
>  pkg/service/repair/worker.go          | 61 +------------------------------------------------------------
>  4 files changed, 49 insertions(+), 128 deletions(-)
> ```

I responded to all your comments, @asias.
Please, let me know if there are other questions/comments.

**Comment by vladzcloudius** [comment]
@tgrabiec your comments were addressed too. Please, re-review.

File: pkg/service/repair/controller.go
Comment (asias) [reply pkg/service/repair/controller.go:202]: Yes, it is existing issue and is outside the scope of the PR.

File: pkg/service/repair/controller.go
Comment (asias) [reply pkg/service/repair/controller.go:201]: No, this does not bother me much and I was not mentioning this for any perf reasons. I just want to make sure I understand the code correctly (the plus one thing).

File: pkg/service/repair/controller.go
Comment (asias) [reply pkg/service/repair/controller.go:193]: Yes, this is the most important change in this PR.

**Review by asias — APPROVED** [review]

**Review by tarzanek — APPROVED** [review]

**Comment by Michal-Leszczynski** [comment]
@vladzcloudius can we close this PR?

**Pull request closed without merging**
